clean_dockerhub.py

```diff
@@ -0,0 +1,277 @@
+import asyncio
+import re
+import time
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+
+import aiohttp
+from rich.console import Console
+from tqdm import tqdm
+
+DOCKERHUB_API = "https://hub.docker.com/v2"
+
+
+class TooManyRequestsError(Exception):
+    """Raised when a 429 Too Many Requests is returned by Docker Hub."""
+
+
+@dataclass
+class Tag:
+    name: str
+    date: datetime
+    deleted: bool
+
+
+async def get_jwt_token(
+    session: aiohttp.ClientSession, username: str, password: str
+) -> str:
+    """
+    Log in to Docker Hub and retrieve a JWT token.
+    """
+    login_url = f"{DOCKERHUB_API}/users/login/"
+    payload = {"username": username, "password": password}
+
+    async with session.post(login_url, json=payload) as resp:
+        resp.raise_for_status()
+        data = await resp.json()
+        return data["token"]
+
+
+def parse_iso_datetime(dt_str: str) -> datetime:
+    """
+    Parse an ISO8601 datetime string (with optional 'Z') into a timezone-aware datetime.
+    """
+    if dt_str.endswith("Z"):
+        dt_str = dt_str[:-1]
+    # remove microseconds as datetime.fromisoformat() always expects a fixed number of digits of microseconds,
+    # whereas the ISO string from Dockerhub returns an arbitrary number of digits.
+    pattern = re.compile(r"\.\d+$")
+    dt_str = pattern.sub("", dt_str)
+    return datetime.fromisoformat(dt_str)
+
+
+async def fetch_old_tags(
+    session: aiohttp.ClientSession,
+    namespace: str,
+    repo: str,
+    threshold: datetime,
+    token: str,
+    max_number_of_pages: int,
+) -> list[Tag]:
+    """
+    Fetch all tags for a repository, handling pagination.
+    Returns a list of `Tag` objects.
+    """
+    tags: list[Tag] = []
+    page = 1
+    page_size = 100
+
+    max_tags = page_size * max_number_of_pages
+
+    console = Console()
+    status_msg = f"Fetching pages for {namespace}/{repo}...{{n_pages}} pages."
+    with console.status(status_msg.format(n_pages=page)) as status:
+        while True:
+            url = (
+                f"{DOCKERHUB_API}/repositories/"
+                f"{namespace}/{repo}/tags?page={page}&page_size={page_size}"
+            )
+            headers = {"Authorization": f"JWT {token}"}
+            async with session.get(url, headers=headers) as resp:
+                if resp.status == 429:
+                    console.log("[bold yellow]Ran into rate limit.")
+                    # If we run into rate limit here, we process all the tags which we fetched until now.
+                    break
+                resp.raise_for_status()
+                data = await resp.json()
+
+            results = data.get("results", [])
+            if not results:
+                break
+
+            for t in results:
+                # Filter tags older than threshold
+                updated = parse_iso_datetime(t["last_updated"])
+                if updated < threshold:
```

Comment:
please correct me if I'm wrong: 

`updated < threshold` is a date that is _older_ than the threshold, eg. June 2021 is < May 2025.
While your comment claims to "filter tags older than threshold".
Please also note the ambiguous meaning of "filter". I would prefer "ignore" or "keep".